common/src/main/java/com/microsoft/identity/common/internal/fido/FidoChallenge.kt

```diff
@@ -0,0 +1,74 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.fido
+
+import com.microsoft.identity.common.internal.fido.FidoChallengeField.Companion.throwIfInvalidOptionalListParameter
+import com.microsoft.identity.common.internal.fido.FidoChallengeField.Companion.throwIfInvalidProtocolVersion
+import com.microsoft.identity.common.internal.fido.FidoChallengeField.Companion.throwIfInvalidRequiredParameter
+import com.microsoft.identity.common.internal.fido.FidoChallengeField.Companion.throwIfInvalidSubmitUrl
+import com.microsoft.identity.common.java.util.UrlUtil
+
+/**
+ * An object representing a FIDO challenge request from the server.
+ *
+ * @param challenge                 Random data string generated by the server. This is a required parameter.
+ * @param relyingPartyIdentifier    The domain name of the identity provider. This is a required parameter.
+ * @param userVerificationPolicy    Relying party's user verification preferences (required, preferred, none. For AAD, this should always be "required").
+ *                                  This is a required parameter.
+ * @param version                   Passkey Auth protocol version. This is a required parameter.
+ * @param submitUrl                 The Url to which the client submits the response to the server's challenge. This is a required parameter.
+ * @param context                   Server state that needs to be maintained between challenge and response. This is a required parameter.
+ * @param keyTypes                  Array of allowed key types. This is an option auth parameter.
+ * @param allowedCredentials        List of credential IDs the client will use for filtering. This is an option auth parameter.
+ *
+ */
+data class FidoChallenge(
+    val challenge: FidoChallengeField<String>,
+    val relyingPartyIdentifier: FidoChallengeField<String>,
+    val userVerificationPolicy: FidoChallengeField<String>,
+    val version: FidoChallengeField<String>,
+    val submitUrl: FidoChallengeField<String>,
+    val context: FidoChallengeField<String>,
+    val keyTypes: FidoChallengeField<List<String>?>,
+    val allowedCredentials: FidoChallengeField<List<String>?>
+) {
+    companion object {
+        const val DELIMITER = ","
+        const val DEFAULT_USER_VERIFICATION_POLICY = "required"
+
+        @JvmStatic
+        operator fun invoke(redirectUri: String): FidoChallenge {
```

Comment:
Does this method invoke the uri as well, or just parsing and getting the FidoChallenge components?
It looks like latter, should update the method name.

Reply:
It's the latter, but invoke in this case refers to this Kotlin operator: https://kotlinlang.org/docs/operator-overloading.html#invoke-operator. I was trying to create another constructor that could do some preprocessing (which a secondary operator would not allow), but unfortunately the invoke() needs to be mentioned when calling from java.
Changed from an operator to a static fun createFromRedirectUri, for better clarity: a376703a0ab288a577766be373400e1e0628536f